core/trino-main/src/main/java/io/trino/util/StatementUtils.java

```diff
@@ -327,4 +381,21 @@ public Class<? extends DataDefinitionTask<T>> getTaskType()
             return taskType.orElseThrow(() -> new IllegalStateException(queryType + " does not have a task type"));
         }
     }
+
+    private static HashSet<Table> getTables(io.trino.sql.tree.Node node, HashSet<String> withTablesToExclude)
+    {
+        HashSet<Table> queryTables = new HashSet<Table>();
+
+        node.getChildren().forEach(child -> {
+            if (child instanceof WithQuery) {
+                withTablesToExclude.add(((WithQuery) child).getName().toString());
+            }
+            else if (child instanceof Table && !withTablesToExclude.contains(((Table) child).getName().toString())) {
```

Comment:
Can you check what polymorphic table functions are represented as in a `Statement` and handle those? They should not be counted as metadata queries IMO.